core/src/main/java/com/graphhopper/storage/index/Snap.java

```diff
@@ -145,12 +146,31 @@ public void calcSnappedPoint(DistanceCalc distCalc) {
         double queryLat = getQueryPoint().lat, queryLon = getQueryPoint().lon;
         double adjLat = fullPL.getLat(wayIndex + 1), adjLon = fullPL.getLon(wayIndex + 1);
         if (distCalc.validEdgeDistance(queryLat, queryLon, tmpLat, tmpLon, adjLat, adjLon)) {
-            GHPoint tmpPoint = distCalc.calcCrossingPointToEdge(queryLat, queryLon, tmpLat, tmpLon, adjLat, adjLon);
+            GHPoint crossingPoint = distCalc.calcCrossingPointToEdge(queryLat, queryLon, tmpLat, tmpLon, adjLat, adjLon);
             double adjEle = fullPL.getEle(wayIndex + 1);
-            snappedPoint = new GHPoint3D(tmpPoint.lat, tmpPoint.lon, (tmpEle + adjEle) / 2);
-        } else
-            // outside of edge boundaries
-            snappedPoint = new GHPoint3D(tmpLat, tmpLon, tmpEle);
+
+            // We want to prevent extra virtual nodes and very short virtual edges in case the snap/crossing point is
+            // very close to a tower node. Since we delayed the calculation of the crossing point until here, we need
+            // to correct the Snap.Position in these cases. Note that it is possible that the query point is very far
+            // from the tower node, but the crossing point is still very close to it.
+            if (considerEqual(crossingPoint.lat, crossingPoint.lon, tmpLat, tmpLon)) {
+                snappedPosition = wayIndex == 0 ? Position.TOWER : Position.PILLAR;
+                snappedPoint = new GHPoint3D(tmpLat, tmpLon, tmpEle);
+            } else if (considerEqual(crossingPoint.lat, crossingPoint.lon, adjLat, adjLon)) {
+                wayIndex++;
+                snappedPosition = wayIndex == fullPL.size() - 1 ? Position.TOWER : Position.PILLAR;
+                snappedPoint = new GHPoint3D(adjLat, adjLon, adjEle);
+            } else {
+                snappedPoint = new GHPoint3D(crossingPoint.lat, crossingPoint.lon, (tmpEle + adjEle) / 2);
+            }
+        } else {
+            // outside of edge segment [wayIndex, wayIndex+1] should not happen for EDGE
+            assert false : "incorrect pos: " + snappedPosition + " for " + snappedPoint + ", " + fullPL + ", " + wayIndex;
+        }
+    }
+
+    public static boolean considerEqual(double lat, double lon, double lat2, double lon2) {
+        return Math.abs(lat - lat2) < 1e-6 && Math.abs(lon - lon2) < 1e-6;
```

Comment:
Btw for example this:

```java
System.out.println(DIST_PLANE.calcDist(49.600000, 11.600000, 49.600001, 11.600001));
```

returns 13cm(!), so not sure if this threshold makes a lot of sense: We consider points to be equal when they are within 13cm from each other, but when they are 20cm from each other we think they are different? Maybe this would be appropriate for indoor routing, but if this tweak is important the threshold should rather be in the ~10m range for road networks?

Reply:
If the granularity (or very short edges) did not matter, then we would not need this code at all :) . If it is important to prevent extra virtual nodes / very short virtual edges from being created I think the `1.e-6` threshold is much too small: If we tolerate 20cm difference why do we think a 13cm difference is problematic? They are still 'same' points for all practical applications.

Anyway, since currently we are not aware of any real problems related to this, we can just leave it as it is as far as I am concerned.